src/System.Diagnostics.TraceSource/src/System/Diagnostics/CorrelationManager.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    public class CorrelationManager
+    {
+        internal CorrelationManager()
+        {
+        }
+
+        private readonly AsyncLocal<Stack> _idStack = new AsyncLocal<Stack>();
```

Comment:
I think we had an issue on Desktop where the CorrelationManager got confused across await calls.

I might be understanding this wrong, but I believe we should get the same stack across async calls so that we can correlate start/stop across awaits.

As for multiple threads - I thought you get one CallContext per thread and then the same CorrelationManager would 'stick' to the thread making multiple thread access on the same CorrelationManager not be an issue.